apps/base-docs/docs/building-with-base/guides/onchain-generative-nfts.md

```diff
@@ -0,0 +1,876 @@
+```
+
+You can use one of the many tools online to do this. Do so, and you'll get:
+
+```text
+{"name": "Land, Sea, and Sky #: 1", "description": "Land, Sea, and Sky is a collection of generative art pieces stored entirely onchain.", "image": "data:image/svg+xml;base64,TODO: Build the SVG with the token ID as the seed"}
+```
+
+## Building the SVG
+
+Next, you need to build logic to compile a real, working svg from the pieces you've saved. You'll also need to add some variation based on the ID of the NFT.
+
+### SVG Renderer Contract
+
+Add a new file and contract called `SVGRenderer`. It doesn't need a constructor, but it will need the `Strings` library:
+
+```solidity
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.20;
+
+import "hardhat/console.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+contract SVGRenderer {
+
+}
+```
+
+Open the exemplar svg in a code editor, and using it as an example, build out a function that uses `abi.encodePacked` to build everything from the svg **except** the actual art. That's much to big for one contract, so add stubs instead.
+
+Depending on the tool you used to make the svg, there may be unneeded extras you can remove from these lines. You also **don't** need the items in `<defs>` or `<styles>`. You'll take advantage of the flexibility of the format to include those in the pieces returned by the supporting contract.
+
+```solidity
+function render(uint _tokenId) public view returns (string memory) {
+  return string(
+    abi.encodePacked(
+      "<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 1024 1024'>",
+      // TODO: Add the clouds,
+      // TODO: Add the sun,
+      // TODO: Add the land,
+      // TODO: Add the sea,
+      // TODO: Add the background,
+      "</svg>"
+    )
+  );
+}
+```
+
+### Rendering the Sea
+
+The sea element of this NFT will be the same for all NFTs, so it makes sense to write that contract first. Create it called, `SeaRenderer`, with a function called `render`. The `<g>` element is the root of the different pieces of the svg, so add that and a stub for the rest.
+
+```solidity
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.20;
+
+import "hardhat/console.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+contract SeaRenderer {
+  function render() public pure returns (string memory) {
+    return // TODO: Render the sea
+  }
+}
+```
+
+The next part is going to be a little tricky, and a little messy. You'll need to combine parts of the individually exported svg that has the sea art and all of its properties with the position data for this part of the art from the exemplar svg. You'll then need to flatten it to a single line, and add it as a string constant.
```

Comment:
Don't need "going" - just say the next part is a little tricky, and a little messy